app/uk/gov/hmrc/agentsubscriptionfrontend/controllers/VatDetailsController.scala

```diff
@@ -56,23 +57,41 @@ class VatDetailsController @Inject()(
     with SessionBehaviour {
 
   def showRegisteredForVatForm: Action[AnyContent] = Action.async { implicit request =>
-    withSubscribingAgent { _ =>
+    withSubscribingAgent { agent =>
       sessionStoreService.fetchAgentSession.flatMap {
         case Some(agentSession) =>
-          (agentSession.businessType, agentSession.registeredForVat) match {
-            case (Some(businessType), Some(registeredForVat)) =>
-              val rfv = RegisteredForVat(RadioInputAnswer.apply(registeredForVat.toString))
-              Ok(html.registered_for_vat(registeredForVatForm.fill(rfv), getBackLink(businessType)))
-            case (Some(businessType), None) =>
-              Ok(html.registered_for_vat(registeredForVatForm, getBackLink(businessType)))
-
-            case _ => Redirect(routes.BusinessTypeController.showBusinessTypeForm())
+          withAAChecks(agent, agentSession) {
+            (agentSession.businessType, agentSession.registeredForVat) match {
+              case (Some(businessType), Some(registeredForVat)) =>
+                val rfv = RegisteredForVat(RadioInputAnswer.apply(registeredForVat.toString))
+                Ok(html.registered_for_vat(registeredForVatForm.fill(rfv), getBackLink(businessType)))
+              case (Some(businessType), None) =>
+                Ok(html.registered_for_vat(registeredForVatForm, getBackLink(businessType)))
+
+              case _ => Redirect(routes.BusinessTypeController.showBusinessTypeForm())
+            }
           }
         case None => Redirect(routes.BusinessTypeController.showBusinessTypeForm())
       }
     }
   }
 
+  private def withAAChecks(agent: Agent, existingSession: AgentSession)(result: Result): Result =
```

Comment:
This looks like a very similar structure to the "showDateOfBirthForm" code above in DateOfBirthController.  Can the two be merged into a single method that implements this logic?  Is the logic really the same?

Reply:
yeah, I checked the code duplication think, but I think we don't get much benefit there with a single method, I can explain why when you are in the office